JuliaLowering/src/scope_analysis.jl

```diff
@@ -17,493 +17,405 @@ function NameKey(ex::SyntaxTree)
-                     selfarg.var_id : var_ids[NameKey(selfarg)]
-            LambdaBindings(selfid)
+# Produce a complete ScopeInfo and add it to the stack of active scopes.  This
+# means finding all variables declared and used in the scope `ex` and generating
+# the (identifier,layer)=>binding_id mapping `scope.vars`
+function enter_scope!(ctx, ex)
+    @assert kind(ex) in KSet"lambda scope_block"
+    # Note that generated functions produce lambdas with this false
+    is_toplevel_thunk = kind(ex) === K"lambda" && ex.is_toplevel_thunk
+    parent_id = (is_toplevel_thunk || isempty(ctx.scope_stack)) ?
+        0 : ctx.scopes[ctx.scope_stack[end]].id
+    is_permeable = is_toplevel_thunk ||
+        kind(ex) === K"scope_block" && ex.scope_type === :neutral
+    scope = ScopeInfo(ctx, parent_id, ex._id, kind(ex) === K"lambda", is_permeable)
+    lambda_scope = ctx.scopes[scope.lambda_id]
+    push!(ctx.scope_stack, scope.id)
+
+    #---------------------------------------------------------------------------
+    # Find explicit decls that may influence assignment assignment resolution
+    if kind(ex) === K"lambda"
+        for c in children(ex[1])
+            @assert kind(c) in KSet"Identifier BindingId Placeholder"
+            maybe_declare_in_scope!(ctx, scope, c, :argument)
+        end
+        for c in children(ex[2])
+            @assert kind(c) in KSet"Identifier BindingId Placeholder"
+            maybe_declare_in_scope!(ctx, scope, c, :static_parameter)
+        end
+        for c in children(ex)[3:end]
+            _find_scope_decls!(ctx, scope, c)
         end
     else
-        parentscope.lambda_bindings
-    end
-
-    for id in values(var_ids)
-        binfo = lookup_binding(ctx, id)
-        if !binfo.is_ssa && binfo.kind !== :global
-            init_lambda_binding(lambda_bindings, id)
+        for c in children(ex)
+            _find_scope_decls!(ctx, scope, c)
         end
     end
 
-    # FIXME: This assumes used bindings are internal to the lambda and cannot
-    # be from the environment, and also assumes they are assigned. That's
-    # correct for now but in general we should go by the same code path that
-    # identifiers do.
-    for id in used_bindings
-        binfo = lookup_binding(ctx, id)
-        if (binfo.kind === :local && !binfo.is_ssa) || binfo.kind === :argument ||
-                binfo.kind === :static_parameter
-            if !has_lambda_binding(lambda_bindings, id)
-                init_lambda_binding(lambda_bindings, id)
-            end
-        end
+    #---------------------------------------------------------------------------
+    # Find assignment targets, possibly introducing implicit locals and globals
+    for (bid, node_id) in sort!(collect(scope.binding_assignments))
+        # mutable nameless bindings may be introduced in desugaring
+        maybe_declare_in_scope!(ctx, scope, SyntaxTree(ctx.graph, node_id),
+                                get_binding(ctx, bid).kind)
     end
```

Comment:
It feels like it should be possible to unify this with the logic for `scope.assignments` below, since semantically they aren't any different iiuc?

Not a required re-work by any means, but it seems like it might be nice